homeassistant/components/todo/__init__.py

```diff
@@ -0,0 +1,312 @@
+"""The todo integration."""
+
+import dataclasses
+import datetime
+import logging
+from typing import Any
+
+import voluptuous as vol
+
+from homeassistant.components import websocket_api
+from homeassistant.components.websocket_api import ERR_NOT_FOUND, ERR_NOT_SUPPORTED
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_ENTITY_ID
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.config_validation import (  # noqa: F401
+    PLATFORM_SCHEMA,
+    PLATFORM_SCHEMA_BASE,
+)
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.entity_component import EntityComponent
+from homeassistant.helpers.typing import ConfigType
+
+from .const import DOMAIN, TodoItemStatus, TodoListEntityFeature
+
+_LOGGER = logging.getLogger(__name__)
+
+SCAN_INTERVAL = datetime.timedelta(seconds=60)
+
+ENTITY_ID_FORMAT = DOMAIN + ".{}"
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up Todo entities."""
+    component = hass.data[DOMAIN] = EntityComponent[TodoListEntity](
+        _LOGGER, DOMAIN, hass, SCAN_INTERVAL
+    )
+
+    websocket_api.async_register_command(hass, websocket_handle_todo_item_list)
+    websocket_api.async_register_command(hass, websocket_handle_todo_item_create)
+    websocket_api.async_register_command(hass, websocket_handle_todo_item_delete)
+    websocket_api.async_register_command(hass, websocket_handle_todo_item_update)
+    websocket_api.async_register_command(hass, websocket_handle_todo_item_move)
```

Comment:
Shouldn't we also register services, besides websocket commands, for all the actions?

Reply:
Or maybe only do services? What's the reason for the websocket command instead of service?

Reply:
> What's the reason for the websocket command instead of service?

I wouldn't say *instead of*, its just limiting scope for initial PR to include what frontend needs (e.g. what  https://github.com/home-assistant/frontend/pull/17883 used though that will be replaced). Happy to change or increase scope here. Part of this may be a misunderstanding (see net). 

> The other four ones,... I would do as services

OK so:
- We prefer reusing services in FE calls whenever possible instead of dedicated websocket calls? My impression was the design standard was that we duplicate websocket APIs and services calls, based on everything I had previously seen, but that may just be inferring somethign incorrect from the legacy code base.   I am definitely all for not duplicating websockets and services, it just has not been a standard I've seen anywhere.
- Given that, What is the distinction on read vs write duplicating functionality in websockets and services? 

Regarding list items as a service: Are we aligned on https://github.com/home-assistant/architecture/discussions/948#discussioncomment-7197201 to move forward yet? 



Aside: One reason for avoiding services was lack of movement on the service call responses discussion (we're still debating details 1+ year out so have been looking to see more alignment there). Since I first created this in the mean time there has been more discussion, though would still like to see a decision to move forward and next steps are opaque.

Reply:
The actions that change state don't return anything currently so I don't think the frontend needs a websocket command. There's no data to return and no subscription needed. A service call should be fine for those.

For the list items action we need to return data. The frontend normally wants a websocket command in that case is my experience. A service would be nice to leverage in automations but may not be needed initially.

The frontend team should confirm this.

Reply:
Thank you for the detail, got it. I've updated this to use services for the mutating actions, preserving the list websocket. I confirmed with another opinion that services are OK.  All tests were rewritten to use the services instead of websockets.

We'll need to add the list service with responses before the services will be useful outside the frontend, but that should be feasible to do.

Reply:
(I'll want to add service response values for the create item service as well)

Reply:
These action apis were made for frontend but I'm wondering if they'll be usable by end users.

Shopping list was designed for matching tasks by name, but that is not how the various cloud task apis work. However, given home assistant make assumptions about finite task list limits, we could still make it work by name without ids. As in, delete takes names instead of IDs.


(Relatedly ,we could enforce the limits in the entity e.g. no more than 500 items in the Todo list etc)

Reply:
Sounds worth to explore.